edb/server/protocol/auth_ext/otc.py

```diff
@@ -0,0 +1,238 @@
+            "code_hash": code_hash,
+            "expires_at": expires_at.isoformat(),
+        },
+        cached_globally=True,
+        query_tag='gel/auth',
+    )
+
+    result_json = json.loads(r.decode())
+    if len(result_json) != 1:
+        raise errors.InvalidData("Failed to create OneTimeCode")
+
+    return code, uuid.UUID(result_json[0]["id"])
+
+
+async def verify(db: Any, factor_id: str, code: str) -> str:
+    """
+    Verify a one-time code for a given factor.
+
+    This function performs all verification operations in a single atomic query:
+    - Cleanup expired codes
+    - Check rate limits
+    - Find and validate the code
+    - Delete the code if valid
+    - Record the authentication attempt
+
+    Args:
+        db: Database connection
+        factor_id: The ID of the factor
+        code: The code to verify
+
+    Returns:
+        The OneTimeCode ID if verification succeeds
+
+    Raises:
+        NoIdentityFound: If code is invalid, expired, or max attempts exceeded
+    """
+    code_hash = hash_code(code)
+
+    # Single composed query that handles the entire verification workflow
+    r = await execute.parse_execute_json(
+        db=db,
+        query="""\
+with
+    factor_id := <uuid>$factor_id,
+    FACTOR := (select ext::auth::Factor filter .id = factor_id),
+    code_hash := <bytes>$code_hash,
+    now := datetime_current(),
+    window_start := now - <duration>'10 minutes',
+
+    # Cleanup expired codes (side effect)
+    cleanup_count := count(
+        delete ext::auth::OneTimeCode filter .expires_at < now
+    ),
+
+    # Check rate limits
+    failed_attempts := (
+        select count(
+            select ext::auth::AuthenticationAttempt
+            filter .factor = FACTOR
+               and .attempt_type =
+                   ext::auth::AuthenticationAttemptType.OneTimeCode
+               and .successful = false
+               and .created_at > window_start
+        )
+    ),
+
+    # Find the OTC
+    otc := (
+        select ext::auth::OneTimeCode
+        filter .factor = FACTOR and .code_hash = code_hash
+        limit 1
+    ),
+
+    is_rate_limited := failed_attempts >= 5,
+    is_code_found := (exists otc) ?? false,
+    is_code_expired := (otc.expires_at < now) ?? false,
+    is_code_valid := (
+        (is_code_found and not is_code_expired and not is_rate_limited)
+        ?? false
+    ),
+
+    # Delete OTC if valid (side effect)
+    deleted_otc := (
+        delete ext::auth::OneTimeCode
+        filter .id = otc.id and is_code_valid
+    ),
+
+    # Record attempt (side effect)
+    recorded_attempt := (
+        if (exists FACTOR)
+        then (
+            insert ext::auth::AuthenticationAttempt {
+                factor := FACTOR,
+                attempt_type :=
+                    ext::auth::AuthenticationAttemptType.OneTimeCode,
+                successful := is_code_valid,
+            }
+        )
+        else (<ext::auth::AuthenticationAttempt>{})
```

Comment:
Did it require the cast? I at least tried to make `else` blocks not require it on empty sets...

Reply:
Ya know, I didn't actually check, I've just been in the habit of writing casts for empty sets because sometimes I need them, and I don't have a good intuition for when I can omit them. I'll give it a shot!